**Note to grader:** Each question consists of parts, e.g. Q1(i), Q1(ii), etc. Each part must be graded  on a 0-4 scale, following the standard NJIT convention (A:4, B+: 3.5, B:3, C+: 2.5, C: 2, D:1, F:0). 
The total score must be re-scaled to 100 &mdash; that should apply to all future assignments so that Canvas assigns the same weight on all assignments. 



In [ ]:
# Grader's area

import numpy as np
G = np.zeros([10,10])
maxScore = 0

# **Assignment 5**

<font>

In [ ]:
from IPython.display import Math
from IPython.display import Latex

import numpy as np
import os
import struct
import time

from sklearn.metrics import classification_report,confusion_matrix
from sklearn.neural_network import MLPClassifier

----------------
----------------
----------------

## <font color='blue'> Question 0. MLP with scikit-learn: Explore outputs for the MNIST dataset </font>

The purpose of this question is to get some practice time with MLP classifiers. We will be using the the MNIST data set. 




In [ ]:
# Let's first load the MNIST data set. 

from sklearn.datasets import fetch_openml
from sklearn.model_selection import train_test_split

X, y = fetch_openml('mnist_784', version=1, return_X_y=True)
y = y.astype(int)
X = ((X / 255.) - .5) * 2
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=10000, random_state=123, stratify=y)

/usr/local/lib/python3.9/dist-packages/sklearn/datasets/_openml.py:968: FutureWarning: The default value of `parser` will change from `'liac-arff'` to `'auto'` in 1.4. You can set `parser='auto'` to silence this warning. Therefore, an `ImportError` will be raised from 1.4 if the dataset is dense and pandas is not installed. Note that the pandas parser may return different data types. See the Notes Section in fetch_openml's API doc for details.
  warn(


---------------------------------------------------------------

**Q0-0**: This the classifier we built in class. Find a setting for *batch_size* and *max_iter*, so that the subsequent fit function **converges**. 

In [ ]:
mlp_mnist = MLPClassifier(verbose = True, hidden_layer_sizes = (50,), batch_size = 10, \
                          max_iter = 300, solver = 'sgd', activation = 'logistic',\
                          learning_rate = 'constant', learning_rate_init = 0.001, random_state = 1)

In [ ]:
mlp_mnist.fit(X_train,y_train)

Iteration 1, loss = 0.77659209
Iteration 2, loss = 0.35169476
Iteration 3, loss = 0.29370323
Iteration 4, loss = 0.26160018
Iteration 5, loss = 0.23835006
Iteration 6, loss = 0.22020206
Iteration 7, loss = 0.20524977
Iteration 8, loss = 0.19250908
Iteration 9, loss = 0.18166501
Iteration 10, loss = 0.17223005
Iteration 11, loss = 0.16398489
Iteration 12, loss = 0.15674479
Iteration 13, loss = 0.15032197
Iteration 14, loss = 0.14438893
Iteration 15, loss = 0.13925911
Iteration 16, loss = 0.13410892
Iteration 17, loss = 0.12973445
Iteration 18, loss = 0.12544996
Iteration 19, loss = 0.12168089
Iteration 20, loss = 0.11826818
Iteration 21, loss = 0.11486403
Iteration 22, loss = 0.11185901
Iteration 23, loss = 0.10913528
Iteration 24, loss = 0.10600094
Iteration 25, loss = 0.10378532
Iteration 26, loss = 0.10133948
Iteration 27, loss = 0.09901220
Iteration 28, loss = 0.09697084
Iteration 29, loss = 0.09496202
Iteration 30, loss = 0.09295482
Iteration 31, loss = 0.09074511
Iteration 32, los

MLPClassifier(activation='logistic', batch_size=10, hidden_layer_sizes=(50,),
              max_iter=300, random_state=1, solver='sgd', verbose=True)

In [ ]:
predictions = mlp_mnist.predict(X_test)
probability_predictions = mlp_mnist.predict_proba(X_test)

For batch_size = 10, the subsequent fit function converges at 228 iterations.

---------------------------------------------------------------

**Q0-1**: Write a short function that outputs the **second best** prediction for each point &mdash; that is, the label that gets the second highest probability in the softmax output. 

In [ ]:
def Second_Best(p1):
    a = []
    for p in p1:
        a.append(np.argsort(np.array(p))[-2])
    return(a)
print(Second_Best(probability_predictions))

[5, 4, 0, 9, 5, 1, 7, 1, 2, 8, 8, 6, 5, 9, 5, 9, 5, 3, 8, 9, 3, 6, 3, 1, 2, 9, 3, 2, 7, 9, 3, 7, 2, 7, 9, 6, 5, 9, 3, 9, 3, 5, 9, 9, 9, 1, 2, 4, 8, 7, 7, 4, 2, 5, 7, 5, 9, 9, 3, 5, 9, 1, 5, 1, 3, 2, 4, 9, 4, 2, 7, 9, 3, 8, 2, 9, 9, 5, 9, 5, 6, 5, 7, 6, 2, 8, 2, 2, 8, 1, 7, 3, 6, 2, 9, 5, 8, 5, 4, 7, 4, 8, 5, 6, 9, 3, 7, 2, 2, 3, 0, 9, 6, 6, 7, 2, 8, 3, 9, 9, 2, 7, 4, 8, 5, 9, 5, 9, 8, 1, 2, 2, 3, 1, 9, 3, 8, 2, 3, 9, 9, 5, 9, 6, 8, 3, 1, 9, 7, 9, 9, 9, 0, 2, 9, 3, 8, 6, 4, 3, 2, 6, 4, 9, 7, 4, 5, 6, 9, 3, 6, 2, 9, 9, 0, 9, 2, 2, 8, 7, 8, 7, 3, 4, 1, 9, 4, 2, 1, 8, 4, 3, 5, 6, 9, 3, 1, 6, 6, 9, 2, 6, 2, 7, 2, 3, 7, 3, 9, 8, 6, 2, 4, 2, 0, 2, 9, 9, 9, 9, 9, 7, 8, 4, 7, 8, 4, 8, 6, 2, 1, 5, 8, 3, 4, 3, 1, 3, 2, 3, 7, 6, 5, 3, 2, 4, 3, 3, 3, 9, 0, 2, 2, 1, 5, 5, 9, 0, 4, 0, 1, 2, 3, 3, 2, 2, 3, 2, 5, 3, 2, 0, 2, 6, 7, 7, 9, 4, 7, 9, 3, 8, 9, 9, 3, 3, 9, 4, 3, 9, 9, 4, 7, 7, 1, 4, 2, 4, 3, 6, 7, 1, 0, 0, 2, 3, 3, 9, 8, 6, 9, 4, 2, 6, 8, 7, 2, 2, 5, 6, 3, 3, 3, 0, 3, 3, 6, 9, 4, 9, 5, 3, 2, 

---------------------------------------------------------------

**Q0-2**: For each data point misclassified by the model, determine whether the second best prediction is actually the correct label. Calculate the percentage of those misclassified points that are correctly classified based on the second best prediction. 

In [ ]:
a = Second_Best(probability_predictions)
b = y_test.values
Missclassify = []
Original_Missclassify = []
for i in range(0,len(b)):
    if(a[i]==b[i]):
        Missclassify.append(a[i])
        Original_Missclassify.append(b[i])

print('The percentage of those misclassified points is ', len(Missclassify)/len(y_test)*100, '%')

The percentage of those misclassified points is  2.1999999999999997 %


This shows that only 2.2% of the points are missclassified, therefore, the highest probability level is truly matching in almost 98% of the scenarios.

---------------------------------------------------------------

**Q0-3**: From a human point of view, we can perhaps see a frequent similarity between handwritten digits 3 and 5 (sometimes we have to pay closer attention to figure out what the true written digit is). We may want to ask if the MLP also detects digit similarities.    
<br>
Consider then the subset $X_i$ of data points that have label $i$ (e.g. all data points with label 5). For each label $j\neq i$, report the frequency by which label $j$ (e.g. label 3) shows up as the second best prediction for points in $X_i$. 
<br>
<br> 
***Further Clarification***: Suppose we have 100 data points with label i=5. Let's say that for 28 of these 100 points, the second predicted label is j=3. Then, the frequency of j=3 as second-best prediction to i=5 is 28%.  We want to compute that frequency for every i and j.  
<br>
Give your code and comment on your findings.

In [ ]:
x = input ("Input value for first label: ")
y = input ("Input value for second label: ")

First_Prediction = []
Second_Prediction  = []
Counter = 0
Counter_ = 0

for p in probability_predictions:
        First_Prediction.append(np.argsort(np.array(p))[-1])
        Second_Prediction.append(np.argsort(np.array(p))[-2])
for i in range(0, len(First_Prediction)):
    if (First_Prediction[i]==int(x)):
        Counter = Counter + 1
    if (First_Prediction[i]==int(x) and Second_Prediction[i]==int(y)):
        Counter_ = Counter_ + 1

print ("The frequency of label " + str(y) + " as second-best prediction to label " + str(x) + " is: " + str(Counter_/Counter*100) + "%")

Input value for first label: 3
Input value for second label: 5
The frequency of label 5 as second-best prediction to label 3 is: 32.47011952191235%


For this dataset the frequency of j=3 as second-best prediction to i=5 is 32.47%.

---------------------------------------------------------------

In [ ]:
# For grader use only

# insert grade here
# G[0,0] =   
# G[0,1] = 
# G[0,2] =
# G[0,3] = 

maxScore = maxScore + 16

------------------------------
------------------------------
------------------------------

## <font color='blue'>Question 1: Working with warm start for focused training



**Q1-0**. MLP classifiers in scikit-learn have an option for so-called 'warm start'. Do a bit of research to find out what it is. Then:

Give an example of warm start. Specifically
train the network for $k$ epochs, and then using warm start
continue the training for another $k$ epochs (4 points).
Here $k$ is a parameter of your choice.  





In [ ]:
Initial_Time = time.time()

mlp_test = MLPClassifier(verbose = True, hidden_layer_sizes = (50,), batch_size = 10, \
                          max_iter = 5, solver = 'sgd', activation = 'logistic',\
                          learning_rate = 'constant', learning_rate_init = 0.001, random_state = 1, warm_start = False)

mlp_test.fit(X_train,y_train)

Final_Time = time.time()

print(Final_Time - Initial_Time)

Iteration 1, loss = 0.77659209
Iteration 2, loss = 0.35169476
Iteration 3, loss = 0.29370323
Iteration 4, loss = 0.26160018
Iteration 5, loss = 0.23835006
51.73737907409668


/usr/local/lib/python3.9/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:686: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (5) reached and the optimization hasn't converged yet.
  warnings.warn(


In [ ]:
Initial_Time = time.time()

mlp_test = MLPClassifier(verbose = True, hidden_layer_sizes = (50,), batch_size = 10, \
                          max_iter = 5, solver = 'sgd', activation = 'logistic',\
                          learning_rate = 'constant', learning_rate_init = 0.001, random_state = 1, warm_start = True)

mlp_test.fit(X_train,y_train)

Final_Time = time.time()

print(Final_Time - Initial_Time)

Iteration 1, loss = 0.77659209
Iteration 2, loss = 0.35169476
Iteration 3, loss = 0.29370323
Iteration 4, loss = 0.26160018
Iteration 5, loss = 0.23835006
48.08075428009033


/usr/local/lib/python3.9/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:686: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (5) reached and the optimization hasn't converged yet.
  warnings.warn(


When Warm Start is set to True, the machine reuse the solution of the previous call to fit as initialization, otherwise, just erase the previous solution. My guess is that this method should be used for the primary purpose of reducing training time when fitting the same dataset, or when there is no significant concept drift in new data.

In this case, I've initiated the MLP classifier with warm_start = False, and for 5 epochs, it took about 48 seconds. However, with warm_start = True, the training time got reduced slightly.

In [ ]:
Initial_Time = time.time()
mlp_test.fit(X_train,y_train)
Final_Time = time.time()
print(Final_Time - Initial_Time)

Iteration 6, loss = 0.22022162
Iteration 7, loss = 0.20535581
Iteration 8, loss = 0.19261592
Iteration 9, loss = 0.18174624
Iteration 10, loss = 0.17227635
50.50786542892456


Continuing with the  same previously initiated MLP Classifier with warm_start = True, it started from iteration 6, reducing the training time in about 20 seconds.

---------------------------------------------------------------


The training set usually remains the same after a warm start. Here, we will use a very similar (but not identical) dataset after the warm start:
<br>
<br>
**Q1-1**. Randomly divide the MNIST dataset into three parts: *train1* (50000), *train2* (10000), *test* (10000). Make sure that all number classes are present in both *train1* and *train2*. [4 points]


In [ ]:
X, y = fetch_openml('mnist_784', version=1, return_X_y=True)
y = y.astype(int)
X = ((X / 255.) - .5) * 2

X_train_Q1, X_test_Q1, y_train_Q1, y_test_Q1 = train_test_split(X, y, test_size=10000, random_state=123, stratify=y)
X_train1_Q1, X_train2_Q1, y_train1_Q1, y_train2_Q1 = train_test_split(X_train_Q1, y_train_Q1, test_size=10000, random_state=123, stratify=y_train_Q1)

print(len(X_train1_Q1), len(y_train1_Q1), len(X_train2_Q1), len(y_train2_Q1), len(X_test_Q1), len(y_test_Q1))

/usr/local/lib/python3.9/dist-packages/sklearn/datasets/_openml.py:968: FutureWarning: The default value of `parser` will change from `'liac-arff'` to `'auto'` in 1.4. You can set `parser='auto'` to silence this warning. Therefore, an `ImportError` will be raised from 1.4 if the dataset is dense and pandas is not installed. Note that the pandas parser may return different data types. See the Notes Section in fetch_openml's API doc for details.
  warn(


50000 50000 10000 10000 10000 10000


---------------------------------------------------------------

**Q1-2**. Use warm start and train your model using the *train1* dataset for $2k$ epochs. Then, using warm start, continue the training with the *train2* dataset for another $2k$ epochs. [4 points]
<br>
<br>
Here, $k$ is the same parameter that you used above. Notice that each point
is considered $2k$ times in this training, exactly as in Part 0. 
<br>
<br>
In effect, this process of training tries to spend more time learning half of the training set, and then moves to the second part of the training set to spend equal time. 


In [ ]:
Initial_Time = time.time()

mlp_test = MLPClassifier(verbose = True, hidden_layer_sizes = (50,), batch_size = 10, \
                          max_iter = 10, solver = 'sgd', activation = 'logistic',\
                          learning_rate = 'constant', learning_rate_init = 0.001, random_state = 1, warm_start = True)

mlp_test.fit(X_train1_Q1,y_train1_Q1)

Final_Time = time.time()

print(Final_Time - Initial_Time)

Iteration 1, loss = 0.84805406
Iteration 2, loss = 0.37653715
Iteration 3, loss = 0.31228990
Iteration 4, loss = 0.27825214
Iteration 5, loss = 0.25438018
Iteration 6, loss = 0.23578022
Iteration 7, loss = 0.22005485
Iteration 8, loss = 0.20718801
Iteration 9, loss = 0.19564455
Iteration 10, loss = 0.18548536
85.14342975616455


/usr/local/lib/python3.9/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:686: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (10) reached and the optimization hasn't converged yet.
  warnings.warn(


In [ ]:
Initial_Time = time.time()
mlp_test.fit(X_train2_Q1, y_train2_Q1)
Final_Time = time.time()
print(Final_Time - Initial_Time)

Iteration 11, loss = 0.19570795
Iteration 12, loss = 0.18875298
Iteration 13, loss = 0.18331168
Iteration 14, loss = 0.17877808
Iteration 15, loss = 0.17410697
Iteration 16, loss = 0.16985178
Iteration 17, loss = 0.16686383
Iteration 18, loss = 0.16271503
Iteration 19, loss = 0.15976617
Iteration 20, loss = 0.15619614
17.34640884399414


For 2k epochs (1) with the same training dataset, it took 85 seconds, for the next 10 epochs, it took only 17 seconds. Therefore, this strategy works well for retraining models faster without the need to start freshly with the full dataset.

---------------------------------------------------------------

**Q1-3**. Which of the two above strategies
works better (Q1-0 or Q1-2)? Feel free to experiment and report your findings.

In [ ]:
Initial_Time = time.time()

mlp_test = MLPClassifier(verbose = True, hidden_layer_sizes = (50,), batch_size = 10, \
                          max_iter = 20, solver = 'sgd', activation = 'logistic',\
                          learning_rate = 'constant', learning_rate_init = 0.001, random_state = 1)

mlp_test.fit(X_train_Q1,y_train_Q1)

Final_Time = time.time()

print(Final_Time - Initial_Time)

Iteration 1, loss = 0.77659209
Iteration 2, loss = 0.35169476
Iteration 3, loss = 0.29370323
Iteration 4, loss = 0.26160018
Iteration 5, loss = 0.23835006
Iteration 6, loss = 0.22020206
Iteration 7, loss = 0.20524977
Iteration 8, loss = 0.19250908
Iteration 9, loss = 0.18166501
Iteration 10, loss = 0.17223005
Iteration 11, loss = 0.16398489
Iteration 12, loss = 0.15674479
Iteration 13, loss = 0.15032197
Iteration 14, loss = 0.14438893
Iteration 15, loss = 0.13925911
Iteration 16, loss = 0.13410892
Iteration 17, loss = 0.12973445
Iteration 18, loss = 0.12544996
Iteration 19, loss = 0.12168089
Iteration 20, loss = 0.11826818
216.54080843925476


/usr/local/lib/python3.9/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:686: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (20) reached and the optimization hasn't converged yet.
  warnings.warn(


---------------------------------------------------------------

*   Strategy 1: same training dataset trained using warm start and in 2 sets of 5 epochs each (10 epoch in total). 98.58 seconds.
*   Strategy 2: different training datasets: train 1 (50000), train 2(10000) of 5 epochs each (20 epoch in total). 102.49 seconds.
*   Strategy 3: experimenting with 60000 records (20 epoch in total). 216.54 seconds.

From the above strategies, my observation is that option 3 performs the best considering it has 60000 records. Warm start makes it faster to retrain with new incoming dataset, however, the loss is not better than cold start. Also, it's normal for soption 1 to peform less efficiently based on lesser number of epochs.

In [ ]:
# For grader use only

# insert grade here
# G[1,0] =   
# G[1,1] = 
# G[1,2] =
# G[1,3] = 

maxScore = maxScore + 16

--------------------------
--------------------------
--------------------------

##  <font color='blue'> Q2. How Powerful are Neural Networks? 

There is a lot of hype surrounding Neural Networks, and their ability to solve problems. This is of course justified, but we should always be aware of their limitations. 


Recall that every row in the matrix $X$  above comes from an image. Every entry of a row corresponds to a (grayscale) pixel. In particular, the first column of $X$ contains the grayscale values of the first pixels of the images in the data set. 
<br>
<br>
We will now 'hide' the label of the image, inside the image itself. 
Here is one way to do that: Replace the top-left pixel of the image (which is a value from 0 to 255) with the label for that image. For example, if the image
contains number 3, then the first pixel of the image will now be set to 3. (In fact, let's set this to 3/255 because we already divided all pixel values by 255.)

---------------------------------------------------------------

**Q2-0.** Create a new dataset $X1$, where each row is the same as the corresponding row of $X$, except in the first column, which contains the label of the corresponding image, divided by 255. Then split
the set $X1$ into a training set and a test set, as we did with $X$. 




In [ ]:
X, y = fetch_openml('mnist_784', version=1, return_X_y=True)
y = y.astype(int)
X = ((X / 255.) - .5) * 2
print(X)

/usr/local/lib/python3.9/dist-packages/sklearn/datasets/_openml.py:968: FutureWarning: The default value of `parser` will change from `'liac-arff'` to `'auto'` in 1.4. You can set `parser='auto'` to silence this warning. Therefore, an `ImportError` will be raised from 1.4 if the dataset is dense and pandas is not installed. Note that the pandas parser may return different data types. See the Notes Section in fetch_openml's API doc for details.
  warn(


       pixel1  pixel2  pixel3  pixel4  pixel5  pixel6  pixel7  pixel8  pixel9  \
0        -1.0    -1.0    -1.0    -1.0    -1.0    -1.0    -1.0    -1.0    -1.0   
1        -1.0    -1.0    -1.0    -1.0    -1.0    -1.0    -1.0    -1.0    -1.0   
2        -1.0    -1.0    -1.0    -1.0    -1.0    -1.0    -1.0    -1.0    -1.0   
3        -1.0    -1.0    -1.0    -1.0    -1.0    -1.0    -1.0    -1.0    -1.0   
4        -1.0    -1.0    -1.0    -1.0    -1.0    -1.0    -1.0    -1.0    -1.0   
...       ...     ...     ...     ...     ...     ...     ...     ...     ...   
69995    -1.0    -1.0    -1.0    -1.0    -1.0    -1.0    -1.0    -1.0    -1.0   
69996    -1.0    -1.0    -1.0    -1.0    -1.0    -1.0    -1.0    -1.0    -1.0   
69997    -1.0    -1.0    -1.0    -1.0    -1.0    -1.0    -1.0    -1.0    -1.0   
69998    -1.0    -1.0    -1.0    -1.0    -1.0    -1.0    -1.0    -1.0    -1.0   
69999    -1.0    -1.0    -1.0    -1.0    -1.0    -1.0    -1.0    -1.0    -1.0   

       pixel10  ...  pixel7

In [ ]:
X_Q2 = X
X_Q2.pixel1 = y/255
print(X_Q2)

         pixel1  pixel2  pixel3  pixel4  pixel5  pixel6  pixel7  pixel8  \
0      0.019608    -1.0    -1.0    -1.0    -1.0    -1.0    -1.0    -1.0   
1      0.000000    -1.0    -1.0    -1.0    -1.0    -1.0    -1.0    -1.0   
2      0.015686    -1.0    -1.0    -1.0    -1.0    -1.0    -1.0    -1.0   
3      0.003922    -1.0    -1.0    -1.0    -1.0    -1.0    -1.0    -1.0   
4      0.035294    -1.0    -1.0    -1.0    -1.0    -1.0    -1.0    -1.0   
...         ...     ...     ...     ...     ...     ...     ...     ...   
69995  0.007843    -1.0    -1.0    -1.0    -1.0    -1.0    -1.0    -1.0   
69996  0.011765    -1.0    -1.0    -1.0    -1.0    -1.0    -1.0    -1.0   
69997  0.015686    -1.0    -1.0    -1.0    -1.0    -1.0    -1.0    -1.0   
69998  0.019608    -1.0    -1.0    -1.0    -1.0    -1.0    -1.0    -1.0   
69999  0.023529    -1.0    -1.0    -1.0    -1.0    -1.0    -1.0    -1.0   

       pixel9  pixel10  ...  pixel775  pixel776  pixel777  pixel778  pixel779  \
0        -1.0     

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X_Q2, y, test_size=10000, random_state=123, stratify=y)

---------------------------------------------------------------

**Q2-1.** With the label of the image 'planted' in the image itself, a Neural Network can achieve accuracy 100%. This is because all the weights connecting to pixels $2...28^2$ can be set to 0, and so effectively the only feature that will be taken into account is the first pixel. The question is whether an MLP can indeed 'discover' this weighting via training.
<br>
<br>
Use an MLP classifier  with one hidden layer of any size you want. Train it on $X'$ and report the accuracy. Does it reach 100\%? Is the accuracy better relative to training with the original input $X$? 




In [ ]:
MLP_MNIST_Planted = MLPClassifier(verbose = True, hidden_layer_sizes = (5,), batch_size = 10, \
                          max_iter = 20, solver = 'sgd', activation = 'logistic',\
                          learning_rate = 'constant', learning_rate_init = 0.001, random_state = 1)

MLP_MNIST_Planted.fit(X_train,y_train)

Iteration 1, loss = 1.52524964
Iteration 2, loss = 0.88900181
Iteration 3, loss = 0.71956766
Iteration 4, loss = 0.63691140
Iteration 5, loss = 0.58163900
Iteration 6, loss = 0.54010832
Iteration 7, loss = 0.51188181
Iteration 8, loss = 0.49304710
Iteration 9, loss = 0.47996592
Iteration 10, loss = 0.46996303
Iteration 11, loss = 0.46123391
Iteration 12, loss = 0.45447035
Iteration 13, loss = 0.44961772
Iteration 14, loss = 0.44435027
Iteration 15, loss = 0.44083022
Iteration 16, loss = 0.43791645
Iteration 17, loss = 0.43551926
Iteration 18, loss = 0.43194595
Iteration 19, loss = 0.42944450
Iteration 20, loss = 0.42773074


/usr/local/lib/python3.9/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:686: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (20) reached and the optimization hasn't converged yet.
  warnings.warn(


MLPClassifier(activation='logistic', batch_size=10, hidden_layer_sizes=(5,),
              max_iter=20, random_state=1, solver='sgd', verbose=True)

In [ ]:
from sklearn.metrics import confusion_matrix
y_pred = MLP_MNIST_Planted.predict(X_test)
CM = confusion_matrix(y_pred, y_test)
print(CM)

from sklearn.metrics import accuracy_score
accuracy_score(y_pred, y_test)

[[ 889    0   20    8    0    9   13    3    8    5]
 [   0 1080   16    8    2    7    3    4   74    1]
 [   9   10  822   21    0    5   21    4   11    0]
 [  24    7   33  891    2   88    0    6   43   12]
 [   4    4   13    3  882   24   40   10   37   50]
 [  20    6    7   37    0  702   20    0   47    5]
 [  15    0   29    1   22    9  877    0    8    1]
 [  17   12   38   36   13   10    0  964   12  143]
 [   5    4   20   13    1   36    6    2  716    5]
 [   3    2    1    2   53   12    2   49   19  772]]


0.8595

The Confusion Matrix shows an accuracy score of 85.95%, therefore, even if the original labels are planted in the first pixel values, a 100% accuracy is not reached.

In [ ]:
# Accuracy using original data (no implanted label within the image) with same hyperparameters.

X, y = fetch_openml('mnist_784', version=1, return_X_y=True)
y = y.astype(int)
X = ((X / 255.) - .5) * 2
X_train_original, X_test_original, y_train_original, y_test_original = train_test_split( X, y, test_size = 10000, random_state = 123, stratify=y)
mlp_mnist_original = MLPClassifier(verbose = True, hidden_layer_sizes = (5,), batch_size = 10, \
                          max_iter = 20, solver = 'sgd', activation = 'logistic',\
                          learning_rate = 'constant', learning_rate_init = 0.001, random_state = 1)

mlp_mnist_original.fit(X_train_original,y_train_original)

from sklearn.metrics import confusion_matrix
y_pred_original = mlp_mnist_original.predict(X_test_original)
CM_original = confusion_matrix(y_pred_original, y_test_original)
print(CM_original)

from sklearn.metrics import accuracy_score
accuracy_score(y_pred_original, y_test_original)

/usr/local/lib/python3.9/dist-packages/sklearn/datasets/_openml.py:968: FutureWarning: The default value of `parser` will change from `'liac-arff'` to `'auto'` in 1.4. You can set `parser='auto'` to silence this warning. Therefore, an `ImportError` will be raised from 1.4 if the dataset is dense and pandas is not installed. Note that the pandas parser may return different data types. See the Notes Section in fetch_openml's API doc for details.
  warn(


Iteration 1, loss = 1.52494227
Iteration 2, loss = 0.88888380
Iteration 3, loss = 0.71952522
Iteration 4, loss = 0.63688762
Iteration 5, loss = 0.58159252
Iteration 6, loss = 0.54008738
Iteration 7, loss = 0.51191195
Iteration 8, loss = 0.49312076
Iteration 9, loss = 0.48007224
Iteration 10, loss = 0.47009466
Iteration 11, loss = 0.46139037
Iteration 12, loss = 0.45464428
Iteration 13, loss = 0.44981172
Iteration 14, loss = 0.44456725
Iteration 15, loss = 0.44106657
Iteration 16, loss = 0.43816835
Iteration 17, loss = 0.43578643
Iteration 18, loss = 0.43223451
Iteration 19, loss = 0.42975625
Iteration 20, loss = 0.42805514
[[ 889    0   19    8    0    9   13    3    8    5]
 [   0 1080   16    8    2    7    3    4   75    1]
 [   9   10  822   21    0    5   21    4   11    0]
 [  24    7   34  891    2   88    0    6   44   12]
 [   4    4   13    3  882   24   40    9   37   50]
 [  20    6    7   37    0  702   20    0   47    5]
 [  15    0   29    1   22    9  877    0    8    1

/usr/local/lib/python3.9/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:686: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (20) reached and the optimization hasn't converged yet.
  warnings.warn(


0.8593

Based on the above confusion matrices and, I can conclude that original data based prediction accuracy (85.93%) is lower than the data having original label planted in pixel 1 (85.95%). Therefore, MLP CLassifier is not able to predict with 100% accuracy, however, predicted with better accuracy than original data.

---------------------------------------------------------------

In [ ]:
# For grader use only

# Insert grade here
# G[2,0] =   
# G[2,1] = 
# G[2,2] =

maxScore = maxScore + 12

In [ ]:
# For grader use
# Total Grade Calculation

rawScore = np.sum(G)
score = rawScore*100/maxScore